### ISSUE TITLE
ToolButton and Button have inconsistent and unergonomic APIs

### ISSUE THREAD

**Issue by Susurrus** [gtk-rs/gtk#438]
The Gtk+ function definition to `set_label` is consistent across `ToolButton` and `Button`, but it's not the case in gtk-rs. Instead `Button` has `fn set_label(&self, label: &str)` and `ToolButton` has `fn set_label(&self, label: Option<&str>)`.

I'm not certain which is the correct one, though I'd argue that neither of them are. You shouldn't need to write `Some(STRING)` and instead use the Into trait, this would change the `set_label` function to look like:
```rust
fn set_label<L>(&self, label: L) where L: Into<Option<&str>> {}
```

Then you should be able to just write `button.set_label("Open")` quite nicely. I found this API suggestion in the [Elegant Library APIs in Rust](https://scribbles.pascalhertleif.de/elegant-apis-in-rust.html#intooption) document yesterday.

**Comment by GuillaumeGomez** [comment]
This is a really great idea! The difference in gtk-rs comes from the fact that in some cases, you can pass `NULL` and in other you can't. With this, the code will be greatly improved!

@EPashkin: Do you want to do it or do you prefer that I do it?

**Comment by Susurrus** [comment]
There are a large number of functions that could use this treatment (`gtk::Label::new()`, `gtk::Scale::add_mark()`, `gtk::FileChooserDialog::new()` are some in my codebase), is this something that can be changed at the generator level to handle all of these cases or does this need to be done on a case-by-case basis manually?

**Comment by GuillaumeGomez** [comment]
@EPashkin: Then so be it! ;)

**Comment by Susurrus** [comment]
Awesome, that'll be very nice to have. Let me know if you'd like help testing.

**Comment by GuillaumeGomez** [comment]
Oh we will! ;)

**Comment by EPashkin** [comment]
@GuillaumeGomez As I remember `T: Into<Option<&str>>` don't work for `&str`. Although I don't checked it now.

**Comment by GuillaumeGomez** [comment]
Really? Is it a lifetime issue?

**Comment by GuillaumeGomez** [comment]
The following code seems to work for me:

```rust
fn lol<'a, L>(s: L) where L: Into<Option<&'a str>> {
    if let Some(s) = s.into() {
        println!("{}", s);
    } else {
        println!("None");
    }
}

fn main() {
    lol(None);
    lol("test");
}
```

**Comment by EPashkin** [comment]
Good.

**Comment by GuillaumeGomez** [comment]
I won't have time this week-end so I'll try to do it starting monday.

**Comment by Susurrus** [comment]
@GuillaumeGomez Just wanted to ping you on this and see if you've made any progress as you mentioned working on it this week.

**Comment by GuillaumeGomez** [comment]
I was about to push something on gir. Be a little more patient. ;)

**Comment by Susurrus** [comment]
Wasn't trying to push! I just remembered this issue in my queue so thought I'd check in. :-)

**Comment by GuillaumeGomez** [comment]
@Susurrus: The beginning of the pr is [here](https://github.com/gtk-rs/gir/pull/314). There isn't much remaining so I hope it'll get merged before the end of the week.

**Comment by GuillaumeGomez** [comment]
Fixed by #447.
